### ISSUE TITLE
[FR] Add warm_start_from in model_to_estimator

### ISSUE THREAD

**Issue by bhack** [tensorflow/tensorflow#19295]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: N/A
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Docker nightly
- **TensorFlow installed from (source or binary)**: Docker nightly
- **TensorFlow version (use command below)**: Docker nightly
- **Python version**: Docker nightly
- **Bazel version (if compiling from source)**: Docker nightly
- **GCC/Compiler version (if compiling from source)**: Docker nightly
- **CUDA/cuDNN version**: Docker nightly
- **GPU model and memory**:  N/A
- **Exact command to reproduce**: N/A

### Describe the problem
Please add `warm_start_from` parameter in `tf.model_to_estimator`

### Source code / logs
N/A

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by shivaniag** [comment]
@fchollet what are your thoughts on this?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @fchollet, @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bhack** [comment]
We have a duplicate at https://github.com/tensorflow/tensorflow/issues/20057

**Comment by tensorflowbutler** [comment]
Nagging Assignees @fchollet, @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by shivaniag** [comment]
@ispirmustafa adding you as well to this feature request if you have any take on this.

**Comment by bhack** [comment]
See also https://github.com/tensorflow/tensorflow/issues/17950#issuecomment-404707663

**Comment by tanzhenyu** [comment]
What's the use case here? (given that user created the model from keras, no checkpoint exists)

**Comment by bhack** [comment]
@tanzhenyu One classical use case is to bootstrap a new experimental from another one created from keras and converted to estimator. That one generated checkpoints.

**Comment by bobbyAtSperry** [comment]
@tanzhenyu Similar use case to above is to train an estimator from a keras model, then later reload it to do predictions, i.e. `.predict`. 

I believe this is meant to be possible but does not seem to work correctly if you don't do the `.predict` immediately after training. I think that loading through warm start might make this possible. For more info on this use case see my issue #20057 (closed as duplicate to this one) and the stack overflow question it refers to - https://stackoverflow.com/questions/50855256/keras-estimator-model-to-estimator-cannot-warm-start-or-load-previous-checkpoi

**Comment by bhack** [comment]
There is also a related problem on this API on how to handle models with BN. See https://github.com/tensorflow/tensorflow/issues/19903

**Comment by tensorflowbutler** [comment]
Nagging Assignees @fchollet, @shivaniag, @ispirmustafa: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bobbyAtSperry** [comment]
I will be out of the office until 6th August and will have little connection to wifi during this period.
  
 If you have any issues with Elmer running please contact wade.garrett@sperryrail.com and abelether.germa@sperryrail.com
  
 I will be catching up with other issues on my return. 
  
 Bobby

**Comment by tensorflowbutler** [comment]
Nagging Assignees @fchollet, @shivaniag, @ispirmustafa: It has been 35 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Silb78dg** [comment]
Hi,
I don't see anything on tf.1.10 or master. Is there any plan to add the warm_start_from parameter?

Thanks.

**Comment by karmel** [comment]
@tanzhenyu -- my understanding is that currently, we warm start from the keras model itself, is that correct? 

@bhack -- can you clarify how you imagine warm starting working in the context of model_to_estimator? The graph comes from the keras model, and the weights from another directory? Can you instead load weights into the Keras model, as they will travel through the warm estimator path in the current code?

**Comment by bhack** [comment]
@karmel I.e. one user case:

- Design a tf.keras model
- Convert that to estimator
- `train_and_evaluate` with automarted model saving from scratch
- Than you want to bootstrap again a different fork of the model with quite the same code from some checkpoint of the previous one.

**Comment by Silb78dg** [comment]
In our case, we are leveraging tf.dataset for pipelining. So it's easier to train with estimator even if some users are more confortable with keras than estimator model_fn... But then you end up with tensorflow checkpoints.
And yes, right now the workaround is to warm start your keras model before converting to an estimator. But it's possible to do it since tf.1.10 before Keras used to support only h5 file

**Comment by bobbyAtSperry** [comment]
Our situation is 

- keras is our choice for building a model
- estimator is our choice to train and evaluate the model
- at some point in the future we wish to further train our estimator model with new data, or use the model in other ways, without having to retrain from scratch.

Problem is that if we want to further train a model we cannot load up an estimator checkpoint. We would have to go back and start from scratch with an untrained model every time we wanted to use it. Hence wanting to warm start an estimator that has been converted from a keras model.

Is @Silb78dg saying we can now at least warm up the keras model from a estimator checkpoint? That might help as we would then only have to reconvert to estimator rather than do previous training all over again. Would still prefer to warm start the estimator model and not have to reconvert though.

**Comment by karmel** [comment]
There are two options here:

1. Warm start the keras model by calling [model.load_weights](https://www.tensorflow.org/api_docs/python/tf/keras/Model#load_weights) -- this will now work with hdf5 or checkpoints. Then, model_to_estimator. For cases where you want to reuse a keras architecture (which is what I think you are saying, @bhack ), you can wrap the model building in a function to build a new model with the same architecture for each new set of weights.

2. Build the Keras model in an Estimator's model_fn, and use the outputs of the keras model to feed into an Estimator Head or an EstimatorSpec directly. An example of this strategy is the [Official MNIST](https://github.com/tensorflow/models/blob/master/official/mnist/mnist.py#L102), and another with Head briefly here:

```
def model_fn(features, labels, mode):
  model = tf.keras.Sequential([
      tf.keras.layers.Dense(64),
      tf.keras.layers.Dense(40, activation=tf.nn.softmax)
  ])
  
  logits = model(features)
  return tf.contrib.estimator.multi_class_head(...).create_estimator_spec(
      mode=mode, features=features, labels=labels, logits=logits, optimizer=tf.train.AdamOptimizer())

```
Using that model_fn in an Estimator will then allow you to use the Estimator API as-is with warm start settings.

Will either of those work for the use cases above? If yes, and you're willing to share a little more-- which do you prefer, and why?

**Comment by Silb78dg** [comment]
@bobbyAtSperry we are on the situation. And yes I confirm that 
`      model = KerasLoader.load_model_from_json(json_file)
        if warm_start_from is not None:
            model.load_weights(warm_start_from)
        model.compile(loss=loss, optimizer=optimizer, metrics=metrics)
        return tf.keras.estimator.model_to_estimator(keras_model=model, config=run_config)`
works perfectly fine for me.

@karmel I don't have any preferences because I'm not the one who write the model and I need to manage both cases (Keras model and model_fn). That's why I was looking for the warm_start parameter but after all this will not simplify my code.

**Comment by bobbyAtSperry** [comment]
@Silb78dg, @karmel, I'll try out your suggestions and report back. They look like they could provide what I need. May not be until next month when we get back onto developing again.

**Comment by lenlen** [comment]
In the warm start case the option 1 allows us to avoid the model_fn implementation, but in option 2, as for summary, we need to implement it. For the option 2 read my comment https://github.com/tensorflow/tensorflow/issues/21983#issuecomment-418826230.

**Comment by tanzhenyu** [comment]
@lenlen Will option 1 work for you?

**Comment by lenlen** [comment]
It should work, but we could be penalized by the lack of functionalities of `WarmStartSettings` as for example vars_to_warm_start to execute partial warm start. In the option 1 you need to replicate some these functionalities on your own.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tanzhenyu: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tanzhenyu: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lenlen** [comment]
@tanzhenyu I've tried the option 1 with the nightly. I used  `model.load_weights`  with my trained model to use the saved tensorflow checkpoint, but I got the error

```
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/checkpointable/util.py", line 888, in streaming_restore
    "Streaming restore not supported from name-based checkpoints. File a "
```
I think that `model.load_weights` only works if the checkpoint is saved with only weights and not for the whole model.
To  simulate the warm start I have to create the estimator from the old model, `model.save_weights` from it, then create a new model and `model.loads_weights` for weights saved before.
Is there an option to load weights from a saved tensorflow checkpoints, without the double model instantiation?

**Comment by bhack** [comment]
@lenlen Seems that it is supported only for object based checkpoints. So it is currently a limit cause you need to work in Eager mode for training.

**Comment by bhack** [comment]
@tanzhenyu Is it correct that option 1 could be used as "a workaround" only in Eager mode?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tanzhenyu: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tanzhenyu: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tanzhenyu: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bobbyAtSperry** [comment]
@Silb78dg, @karmel, I can confirm that we have successfully implemented @karmel's option 1. So rather than warm starting we just save and load to checkpoints as we previously did but can now use the keras syntax to design our models which become standard TF models in the model_fn in the estimator.

The following is a schematic of how we use the various TF components to train_and_validate and to predict using keras, estimators and tf.data. This may be a useful outline for others:

I specify the model with keras notation but then put it in a tensorflow model function which can then be loaded into and estimator. By taking this approach I can save to, and reload from, checkpoints as with any other tensorflow model. I think this gives the best combination of using the intuitive keras notation, while being able to take advantage of the tensorflow estimators and data tools. The following is an overview of my approach describing the setup of the various tensorflow calls:

1.  Create an estimator:

```
|--estimator: tf.estimator.Estimator
|--config: tf.estimator.RunConfig                         #checkpointPath and saving spec for training
        |--model_fn: tf.estimator.EstimatorSpec
            |--myKerasModel                               #specify model. Doesn't have to be keras.
                    |--keras.models.Model
            |--loss: myLossFunction                       #train_and_eval only
            |--optimizer: myOptimizerFunction             #train_and_eval only
            |--training_hooks:tf.train.SummarySaverHook   #train_and_eval only - for saved diagnostics
            |--evaluation_hooks:tf.train.SummarySaverHook #train_and_eval only - for saved diagnostics
            |--predictions: model(data, training=False)   #predict only
```
2.  Train_and_evaluate:
```
|--tf.estimator.train_and_evaluate
        |--estimator: tf.estimator.Estimator                  #the estimator we created
        |--train_spec: tf.estimator.TrainSpec
                            |--input_fn: tf.data.Dataset      #specify the input function 
                                |--features, labels           #data to use
                                |--batch, shuffle, num_epochs #controls for data
        |--eval_spec: tf.estimator.EvalSpec
                            |--input_fn: tf.data.Dataset      #specify the input function
                                |--features, labels           #data to use
                                |--batch                      #controls for data
                            |--throttle and start_delay       #specify when to start evaluation
```
Note that we could use `tf.estimator.Estimator.train` and `tf.estimator.Estimator.evaluate` but that doesn't allow doing evaluations during the training so we use `tf.estimator.train_and_evaluate` instead.

3.  Predict:
```
|--estimator: tf.estimator.Estimator.predict       #the estimator we created
        |--input_fn: tf.data.Dataset               #specify the input function
            |--features                            #data to use
```

**Comment by tanzhenyu** [comment]
I'm marking this as contribution welcome.
